### PR TITLE
Multi-package instrument support

### PR DESCRIPTION
**Pull request by rstoneback**
# Description

Addresses # 265

Adds support so that pysat may look to multiple python packages for code to support a particular instrument. Currently, only files in pysat.instruments have first class support as Instruments. This feature will make it easier to use instruments not shipped within pysat proper. 

## Type of change

Please delete options that are not relevant.

- New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

Code as written currently still works for heritage pysat Instrument operations. New modules have yet to be developed to fully test this code.

**Test Configuration**:
* MacOS
* 10.14.5

# Checklist:

- [X] Make sure you are merging into the ``develop`` (not ``master``) branch
- [X] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [x] Add a note to ``CHANGELOG.md``, summarizing the changes

### CODE REVIEW

**Comment by asher-pembroke** [comment]
When I checked out `instrument_independence` and ran `nosetests -vs pysat` I got the following:
```console
Failure: FileNotFoundError ([Errno 2] No such file or directory: '/Users/apembrok/.pysat/user_modules.txt') ... ERROR
```
This doesn't happen when I run tests on master.

I'm using Python 3.7.3 from miniconda

**Comment by rstoneback** [comment]
There is a small bit of code in the init file (part of the pull) that should have made an empty file with that name. Only runs first time pysat is run. Must need some work.

On Sep 6, 2019, at 5:51 PM, Asher Pembroke <notifications@github.com<mailto:notifications@github.com>> wrote:


When I checked out instrument_independence and ran nosetests -vs pysat I got the following:

Failure: FileNotFoundError ([Errno 2] No such file or directory: '/Users/apembrok/.pysat/user_modules.txt') ... ERROR

This doesn't happen when I run tests on master.

I'm using Python 3.7.3 from miniconda

**Comment by rstoneback** [comment]
ready for review @jklenzing or @aburrell

**Comment by aburrell** [comment]
I just downloaded this branch and imported it.  Immediately got a `FileNotFoundError`.  It looks like the new file `user_modules.txt` is read before it is created.

```
/opt/local/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/importlib/__init__.py in reload(module)
    167         if spec is None:
    168             raise ModuleNotFoundError(f"spec not found for the module {name!r}", name=name)
--> 169         _bootstrap._exec(spec, module)
    170         # The module may have replaced itself in sys.modules!
    171         return sys.modules[name]

/opt/local/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/importlib/_bootstrap.py in _exec(spec, module)

/opt/local/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/importlib/_bootstrap_external.py in exec_module(self, module)

/opt/local/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/importlib/_bootstrap.py in _call_with_frames_removed(f, *args, **kwds)

~/Programs/Git/pysat/pysat/__init__.py in <module>
     81     # load up stored user modules
     82     user_modules = []
---> 83     with open(os.path.join(pysat_dir, 'user_modules.txt'), 'r') as f:
     84         for _ in f:
     85             if _ != '' and (_ is not None):

FileNotFoundError: [Errno 2] No such file or directory: '/Users/aburrell/.pysat/user_modules.txt'
```

**Comment by aburrell** [comment]
Also, to not hold back bug fixes, we should probably create a `develop-3.0.0` branch that we push this to.  That way we can continue to fix problems that arise and still work towards the 3.0.0 release.

**Comment by rstoneback** [comment]
I'm ok with a develop-3.0 branch. Thanks for the note about the bug. This file issue has actually cropped up before. That means we need a dedicated method to handling this, plus handle a multi-thread environment. I'll make an issue and provide a quick fix here.

**Comment by rstoneback** [comment]
Created issue #309 about the user_modules file creation bug.

**Comment by jklenzing** [comment]
Added `develop-3` branch, based on the current develop branch.

File: pysat/__init__.py [hunk 1]
```diff
@@ -70,10 +70,23 @@
...
+        user_modules = []
+
 else:
     # load up stored data path
     with open(os.path.join(pysat_dir, 'data_path.txt'), 'r') as f:
         data_dir = f.readline()
+    # load up stored user modules
+    user_modules = []
+    modules_file = os.path.join(pysat_dir, 'user_modules.txt')
+    if os.path.exists(modules_file):
```
Comment (rstoneback) [inline pysat/__init__.py]: we need to make a empty user_modules if it doesn't exist

**Comment by rstoneback** [comment]
ready for re-review @aburrell

**Comment by jklenzing** [comment]
Just ran through a test with pysatMissionPlanning.  It's working as documented.  @aburrell, I can't find the problem you mentioned the other day.

**Comment by jklenzing** [comment]
Waiting on #325 to merge.

**Comment by jklenzing** [comment]
OK, all prereqs are merged, and I've fixed the conflicts.  I think this is ready to merge in, unless there's another comment.

**Pull request merged**